internal/provider/data_source_gitlab_project_membership.go

```diff
@@ -0,0 +1,184 @@
+							Computed:    true,
+						},
+						"name": {
+							Description: "The name of the user.",
+							Type:        schema.TypeString,
+							Computed:    true,
+						},
+						"state": {
+							Description: "Whether the user is active or blocked.",
+							Type:        schema.TypeString,
+							Computed:    true,
+						},
+						"avatar_url": {
+							Description: "The avatar URL of the user.",
+							Type:        schema.TypeString,
+							Computed:    true,
+						},
+						"web_url": {
+							Description: "User's website URL.",
+							Type:        schema.TypeString,
+							Computed:    true,
+						},
+						"access_level": {
+							Description: "The level of access to the group.",
+							Type:        schema.TypeString,
+							Computed:    true,
+						},
+						"expires_at": {
+							Description: "Expiration date for the group membership.",
+							Type:        schema.TypeString,
+							Computed:    true,
+						},
+					},
+				},
+			},
+		},
+	}
+})
+
+func dataSourceGitlabProjectMembershipRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*gitlab.Client)
+
+	var project *gitlab.Project
+	var err error
+
+	log.Printf("[INFO] Reading Gitlab project")
+
+	if projectID, ok := d.GetOk("project_id"); ok {
+		// Get project by id
+		project, _, err = client.Projects.GetProject(projectID.(int), nil, gitlab.WithContext(ctx))
+		if err != nil {
+			return diag.FromErr(err)
+		}
+	} else if fullPath, ok := d.GetOk("full_path"); ok {
+		// Get project by full path
+		project, _, err = client.Projects.GetProject(fullPath.(string), nil)
+		if err != nil {
+			return diag.FromErr(err)
+		}
+	} else {
+		return diag.Errorf("one and only one of project_id or full_path must be set")
+	}
+
+	log.Printf("[INFO] Reading Gitlab project memberships")
+
+	// Get project memberships
+	listOptions := &gitlab.ListProjectMembersOptions{
+		ListOptions: gitlab.ListOptions{
+			PerPage: 20,
+			Page:    1,
+		},
+	}
+
+	listMembers := client.ProjectMembers.ListProjectMembers
+	if inherited, ok := d.GetOk("inherited"); ok && inherited.(bool) {
+		listMembers = client.ProjectMembers.ListAllProjectMembers
+	}
+
+	var allPMs []*gitlab.ProjectMember
+	for {
+		pms, resp, err := listMembers(project.ID, listOptions, gitlab.WithContext(ctx))
+		if err != nil {
+			return diag.FromErr(err)
+		}
+
+		allPMs = append(allPMs, pms...)
+
+		if resp.NextPage == 0 {
+			break
+		}
+		listOptions.Page = resp.NextPage
+	}
+
+	d.Set("project_id", project.ID)
+	d.Set("full_path", project.PathWithNamespace)
+
+	if err := d.Set("members", flattenGitlabProjectMembers(d, allPMs)); err != nil {
+		return diag.FromErr(err)
+	}
+
+	var optionsHash strings.Builder
+	optionsHash.WriteString(strconv.Itoa(project.ID))
+
+	if data, ok := d.GetOk("access_level"); ok {
+		optionsHash.WriteString(data.(string))
+	}
+
+	id := schema.HashString(optionsHash.String())
+	d.SetId(fmt.Sprintf("%d", id))
```

Comment:
The id should be set first, so that if something goes wrong there isn't any dangling state and tf knows what belongs to where :)

Reply:
With "set first", do you mean right at the beginning of the function - right after we have the project id - or just before we write the other parts of the hash?

Reply:
Just before writing the other states with `d.Set()`.